lite2/client.go

```diff
@@ -129,26 +144,111 @@ func NewClient(
+		trustingPeriod:                     trustOptions.Period,
+		verificationMode:                   skipping,
+		trustLevel:                         DefaultTrustLevel,
+		primary:                            primary,
+		trustedStore:                       trustedStore,
+		removeNoLongerTrustedHeadersPeriod: defaultRemoveNoLongerTrustedHeadersPeriod,
+		quit:                               make(chan struct{}),
+		logger:                             log.NewNopLogger(),
 	}
 
 	for _, o := range options {
 		o(c)
 	}
 
-	if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+	if err := c.restoreTrustedHeaderAndNextVals(); err != nil {
 		return nil, err
 	}
+	if c.trustedHeader != nil {
+		if err := c.checkTrustedHeaderUsingOptions(trustOptions); err != nil {
+			return nil, err
+		}
+	}
+
+	if c.trustedHeader == nil || c.trustedHeader.Height != trustOptions.Height {
+		if err := c.initializeWithTrustOptions(trustOptions); err != nil {
+			return nil, err
+		}
+	}
+
+	go c.removeNoLongerTrustedHeadersRoutine()
 
 	return c, nil
 }
 
+// Load trustedHeader and trustedNextVals from trustedStore.
+func (c *Client) restoreTrustedHeaderAndNextVals() error {
+	lastHeight, err := c.trustedStore.LastSignedHeaderHeight()
+	if err != nil {
+		return errors.Wrap(err, "can't get last trusted header height")
+	}
+
+	if lastHeight > 0 {
+		trustedHeader, err := c.trustedStore.SignedHeader(lastHeight)
+		if err != nil {
+			return errors.Wrap(err, "can't get last trusted header")
+		}
+
+		trustedNextVals, err := c.trustedStore.ValidatorSet(lastHeight + 1)
+		if err != nil {
+			return errors.Wrap(err, "can't get last trusted next validators")
+		}
+
+		c.trustedHeader = trustedHeader
+		c.trustedNextVals = trustedNextVals
+	}
+
+	return nil
+}
+
+// if options.Height:
+//
+//     1) ahead of trustedHeader.Height => fetch header from primary provider
+//     and check it matches trustedHeader (if not, remove trustedHeader and all
+//     the headers before)
+//     2) equals trustedHeader.Height => check Hash is the same (if not, remove
+//     trustedHeader and all the headers before)
+//     3) behind trustedHeader.Height => remove all the headers between
+//     options.Height and trustedHeader.Height, then check the Hash is the same
+//     (if not, remove trustedHeader and all the headers before)
+//
+// The intuition here is the user always right. I.e. if she decides to reset
+// the light client with an older header, there must be a reason for it.
+func (c *Client) checkTrustedHeaderUsingOptions(options TrustOptions) error {
+	var primaryHash []byte
+	switch {
+	case options.Height > c.trustedHeader.Height:
+		h, err := c.primary.SignedHeader(c.trustedHeader.Height)
+		if err != nil {
+			return err
+		}
+		primaryHash = h.Hash()
+	case options.Height == c.trustedHeader.Height:
+		primaryHash = options.Hash
+	case options.Height < c.trustedHeader.Height:
+		// remove all the headers ( options.Height, trustedHeader.Height ]
```

Comment:
Here is one place I think we might want the user to explicitly confirm that they wanted to do this.